### PR TITLE
[Tensor refactor] Subclass ndarray to allow marking of trainable parameters

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

The current method for determining whether QNode parameters are differentiable or not relies on the function signature; named arguments with defaults are assumed to be non-differentiable, while named arguments with no defaults are assumed to be differentiable.

This is inherited from Autograd, but is too implicit, and not at all intuitive.

**Description of the Change:**

This PR is the first in a series to instead change the parameter handling of QNodes, so that arrays are themselves marked as differentiable or non-differentiable, removing any dependence on the function signature.

This PR includes:

* A new module `pennylane.numpy`, which imports `autograd.numpy`, monkeypatches it, and makes it accessible to the user. This is accessible via `from pennylane import numpy`.

* A new `np.ndarray` subclass `qml.np.tensor`, which adds an additional attribute and keyword argument to NumPy arrays: `requires_grad`. This attribute defines whether a NumPy array should be considered differentiable/trainable, or not. For backwards compatibility, the default value is `true`.

* A function `tensor_wrapper`, which iteratively wraps all functions contained within `np.__dict__`, such that:

  - They accept the `requires_grad` keyword argument, and
  - They output `np.tensor` objects.

  This allows array creation functions such as `np.array`, `np.asarray`, `np.zeros` etc. to create `qml.np.tensor` objects, as well as allowing functions that act on arrays (`np.sin`, `np.vstack` etc) to accept `np.tensor` objects as input.

* A small monkeypatch of Autograd, so that it understands how to differentiate `qml.np.tensor` objects. There does not seem to be a valid method for forcing non-differentiable parameters to result in gradients of `0` or `None` in Autograd. I attempted to use [`getval()`](https://github.com/HIPS/autograd/issues/256), but this seems to break autograd in other un-intended ways. In lieu, currently an exception is raised if the user explicitly tries to differentiate a non-differentiable parameter.

**Benefits:** n/a

**Possible Drawbacks:**

* No changes are made to the existing QNode behaviour yet. Subsequent PRs will modify the `BaseQNode`, followed by the interfaces, to use this new behaviour.

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/numpy/__init__.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+"""
+# pylint: disable=wrong-import-position,wildcard-import,undefined-variable
+from autograd import numpy as _np
+from autograd.numpy import *
+
+from .wrapper import wrap_arrays, extract_tensors
+
+wrap_arrays(_np.__dict__, globals())
+
+# Delete the unwrapped fft, linalg, random modules
```
Comment (co9olguy) [inline pennylane/numpy/__init__.py:26]: What is special about these three submodules that they require special care?

File: pennylane/numpy/wrapper.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        else:
+            tensor_args = list(extract_tensors(args))
+
+            if tensor_args:
+                # Unless the user specifies otherwise, if all tensors in the argument
+                # list are non-trainable, the output is also non-trainable.
+                # NOTE: Use of Python's ``all`` results in an infinite recursion,
+                # and I'm not sure why. Using ``np.all`` works fine.
+                tensor_kwargs["requires_grad"] = not _np.all(
+                    [not i.requires_grad for i in tensor_args]
```
Comment (co9olguy) [inline pennylane/numpy/wrapper.py:68]: how come `not` is used twice here?? does it not work to just do `_np.all([i.requires_grad for i in tensor_args])?

File: pennylane/numpy/wrapper.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    elif isinstance(x, Sequence) and not isinstance(x, (str, bytes)):
+        # If the item is a sequence, recursively look through its
+        # elements for tensors.
+        # NOTE: we choose to branch on Sequence here and not Iterable,
+        # as NumPy arrays are not Sequences.
+        for item in x:
+            yield from extract_tensors(item)
+
+
+def tensor_wrapper(obj):
```
Comment (co9olguy) [inline pennylane/numpy/wrapper.py:42]: I found the name `obj` too obscure. Would we be better suited with a more descriptive name of what types we expect to be used (`func`?, `callable_obj`?)

File: pennylane/numpy/wrapper.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                    [not i.requires_grad for i in tensor_args]
+                )
+
+        # evaluate the original object
+        res = obj(*args, **kwargs)
+
+        if isinstance(res, _np.ndarray):
+            # only if the output of the object is a ndarray,
+            # then convert to a PennyLane tensor
+            res = tensor(res, **tensor_kwargs)
```
Comment (co9olguy) [inline pennylane/numpy/wrapper.py:77]: Trying to understand :thinking: 
So a numpy function that returns a scalar float value won't have it's output marked as differentiable?

File: pennylane/numpy/wrapper.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+        return res
+
+    return _wrapped
+
+
+def wrap_arrays(old, new):
+    """Loop through an objects symbol table,
```
Comment (co9olguy) [inline pennylane/numpy/wrapper.py:85]:
```suggestion
    """Loop through an object's symbol table,

```

File: pennylane/numpy/tensor.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        instead.
+
+    .. warning::
+
+        Tensors should be constructed using standard array construction functions
+        provided as part of PennyLane's NumPy implementation, including
+        ``np.array``, ``np.zeros`` or ``np.empty``.
+
+        The parameters given here refer to a low-level class
+        for instantiating an tensors.
```
Comment (co9olguy) [inline pennylane/numpy/tensor.py:52]:
```suggestion
        for instantiating tensors.

```

File: pennylane/numpy/tensor.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            an array. This includes lists, lists of tuples, tuples, tuples of tuples,
+            tuples of lists and ndarrays.
+        requires_grad (bool): whether the tensor supports differentiation
+
+    **Example**
+
+    The trainability of a tensor can be set on construction via the
+    ``requires_grad`` keyword argument,
+
+    >>> from pennylane import numpy as np
```
Comment (co9olguy) [inline pennylane/numpy/tensor.py:65]: I can't remember, do we recommend using `code-block` in these kind of docstrings?

File: pennylane/numpy/tensor.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+    >>> x = np.array([0, 1, 2], requires_grad=False)
+    >>> y = np.zeros([3], requires_grad=True)
+    >>> np.vstack([x, y])
+    tensor([[0., 1., 2.],
+        [0., 0., 0.]], requires_grad=True)
+    """
+
+    def __new__(cls, input_array, *args, requires_grad=True, **kwargs):
+        obj = _np.array(input_array, *args, **kwargs)
```
Comment (co9olguy) [inline pennylane/numpy/tensor.py:92]: I find the name `obj` both too obscure and too suggestive, if that makes sense :thinking:

File: tests/test_numpy_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+
+        res = np.vstack([arr1, arr2, arr3])
+        assert not res.requires_grad
+
+        # If one of the inputs is trainable, the output always is.
+        arr1.requires_grad = True
+        res = np.vstack([arr1, arr2, arr3])
+        assert res.requires_grad
+
+    def test_wrapped_function_nontrainable_variable_args(self):
```
Comment (co9olguy) [inline tests/test_numpy_wrapper.py:221]: I didn't quite get what this test did that the one above wasn't doing :thinking:

File: tests/test_numpy_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+    def test_non_differentiable_gradient(self):
+        """Test gradient computation with requires_grad=False raises an error"""
+        def cost(x):
+            return np.sum(np.sin(x))
+
+        grad_fn = qml.grad(cost, argnum=[0])
+        arr1 = np.array([0., 1., 2.], requires_grad=False)
+
+        with pytest.raises(np.NonDifferentiableError, match="non-differentiable"):
+            grad_fn(arr1)
```
Comment (co9olguy) [inline tests/test_numpy_wrapper.py:293]: Awesome tests :+1:

**Comment by josh146** [comment]
> Are we just falling back to autograd logic (args/kwargs determines differentiability) at that point?

Not quite, more so that the new logic assumes that `requires_grad=True` by default on all floats and standard numpy arrays. Instead, the explicit existence of `requires_grad=False` actively disables differentiation.

This is the TensorFlow approach (all tensors are watched by default), and required for backwards compatibility with tutorials such as qubit rotation (where the trainable inputs to the QNode are python floats).

File: pennylane/__init__.py [hunk 11]
```diff
@@ -207,7 +203,7 @@ def jacobian(func, argnum):
     # pylint: disable=no-value-for-parameter
     if isinstance(argnum, int):
         return _jacobian(func, argnum)
-    return lambda *args, **kwargs: numpy.stack(
+    return lambda *args, **kwargs: _np.stack(
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/__init__.py:206]: Good question, I can't remember why I made this change (this was back when I was still working on the ADR).

File: pennylane/numpy/__init__.py (hunk 1 above)
Comment (josh146) [reply pennylane/numpy/__init__.py:26]: These are the 3 subpackages of NumPy that are supported by Autograd, so we need to re-wrap them to make sure that they support the new `tensor` subclass. 

I don't 100% like the solution here (deleting `fft`, `linalg`, and `random`, and then importing the wrapped versions), but I think there is subtlety to Python's import system I wasn't truly grasping. No matter how I ordered the imports, or where I did the wrapping, `pennylane.numpy.random` would always simply point to `autograd.numpy.random` from the `from autograd.numpy import *` line.

It's only when I delete those modules from `global()` before importing from the relative path does python properly import the wrapped modules.

File: pennylane/numpy/tensor.py (hunk 7 above)
Comment (josh146) [reply pennylane/numpy/tensor.py:65]: Nope, we recommend `pycon` without `.. code-block::` directives. This is the [guideline followed by NumPy/SciPy](https://numpy.org/devdocs/docs/howto_document.html#docstring-standard) (see point (15)), and is compatible with `doctest` if we decide to test our documentation examples in the future.

File: pennylane/numpy/tensor.py (hunk 8 above)
Comment (josh146) [reply pennylane/numpy/tensor.py:92]: I agree, my only reason for keeping it in is that this is the variable name the NumPy documentation gives it: https://numpy.org/devdocs/user/basics.subclassing.html#simple-example-adding-an-extra-attribute-to-ndarray

Let me know if you have any other suggestions!

File: pennylane/numpy/wrapper.py (hunk 4 above)
Comment (josh146) [reply pennylane/numpy/wrapper.py:77]: Since the `tensor` class is a subclass of `ndarray`, we really only want functions that return `ndarrays` to now return `tensors`. We are really taking advantage of duck typing here --- `tensor` objects quack like arrays, so existing workflows will continue to work as follows.

Making _all_ NumPy functions return `tensor` objects I feel would be:

* At best slightly confusing, since `pennylane.numpy` would no longer follow exactly the vanilla NumPy API, users would have to include additional calls to `np.item()` or `np.asscalar()` where needed, and

* At worst would create edge cases and regressions.

Note that this logic here doesn't affect differentiability, since we made the decision in the ADR that all Python floats would be treated as differentiable by default (so the Autograd interface treats `x=2.0` as if `requires_grad=True`). This decision was made for backwards compatibility with demos such as qubit rotation, but we can re-evaluate down the line if backwards-compatibility is no longer an issue.

File: tests/test_numpy_wrapper.py (hunk 9 above)
Comment (josh146) [reply tests/test_numpy_wrapper.py:221]: Ah, I added in this test last minute, as I wanted to check if the wrapping worked for NumPy functions of the form `np.func([arr1, arr2, ...])`, as well as `np.func(arr1, arr2)` 😆 Essentially, just wanted to make sure all bases were covered. Will make this clearer in the docstring.

File: pennylane/numpy/wrapper.py (hunk 2 above)
Comment (josh146) [reply pennylane/numpy/wrapper.py:68]: Oh good question 😆 boolean logic has always confused me

File: pennylane/__init__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/__init__.py:206]: Oh, I remember the motivation behind this change. Essentially, `pennylane.__init__` shouldn't be importing the wrapped version of NumPy, as it places a dependence on Autograd that is not needed. In addition, the output of `qml.jacobian` is not differentiable anyway, since we haven't implemented second derivatives.

File: pennylane/numpy/wrapper.py (hunk 3 above)
Comment (josh146) [reply pennylane/numpy/wrapper.py:42]: I've had a think about this, I don't think `func` or `callable_obj` are general enough, since this wrapper acts on both callables _and_ classes. The only objects it wouldn't explicitly work on are instances of non-callable classes. I'll add this to the docstring.

Comment (josh146) [reply pennylane/numpy/wrapper.py:42]:
```rest
Decorator that wraps callable objects and classes so that they both accept
a ``requires_grad`` keyword argument, as well as returning a PennyLane
:class:`~.tensor`.

Only if the decorated object returns an ``ndarray`` is the
output converted to a :class:`~.tensor`; this avoids superfluous conversion
of scalars and other native-Python types.

Args:
    obj: a callable object or class
```

File: pennylane/numpy/wrapper.py (hunk 2 above)
Comment (josh146) [reply pennylane/numpy/wrapper.py:68]: Have updated this to
```python
_np.any([i.requires_grad for i in tensor_args])
```

(the two `not`s turn the all into an any)

File: tests/test_numpy_wrapper.py (hunk 9 above)
Comment (josh146) [reply tests/test_numpy_wrapper.py:221]: Have updated the docstrings and the test names

**Comment by josh146** [comment]
@co9olguy, have made the requested changes and answered some of your questions; not sure if you would like to re-review or not.

**Comment by co9olguy** [comment]
:shipit:
